### ISSUE TITLE
Type System.Management.Automation.PSMemberInfoIntegratingCollection returned by .psobject.properties should have a .Count property

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#9155]
Accessing `.psobject.properties` on any object is a convenient way to reflect on that object's properties.

Sometimes all you want is to _count_ the number of properties.

That is currently complicated by the `[System.Management.Automation.PSMemberInfoIntegratingCollection[System.Management.Automation.PSPropertyInfo]]` instance returned not implementing a `.Count` property.

The workaround is to wrap the `.psobject.properties` access in `@(...)` before accessing `.Count`, but that shouldn't be necessary.

# Steps to reproduce

```powershell
$obj = [pscustomobject] @{ one = 1; two = 2 }
$obj.psobject.properties.Count
```

# Expected behavior

```none
2
```

# Actual behavior

```none
1
1
```

That is, in the absence of a `.Count` property on the _collection_, member enumeration was used to look for a `.Count` property _on each element_  of the collection, which - due to PSv3+ implementing `.Count` even on _scalars_ - returned `1` for each property.

# Environment data

```none
PowerShell Core 6.2.0-rc.1
Windows PowerShell v5.1 
```

**Comment by iSazonov** [comment]
Actually it is not Collection - it is IEnumerable.
```c#
internal class PSMemberInfoIntegratingCollection<T> : PSMemberInfoCollection<T>, IEnumerable<T> where T : PSMemberInfo

    public abstract class PSMemberInfoCollection<T> : IEnumerable<T> where T : PSMemberInfo
``

**Comment by mklement0** [comment]
Yes, @iSazonov, but given its purpose - and, not least, its _name_, <pre>PSMemberInfoIntegrating<i>Collection</i></pre>  - it's reasonable to expect a `.Count` property, whether as a direct property or via an interface.

**Comment by iSazonov** [comment]
We could implement tha class based on Collection. I am afraid that it is too expensive to use in PSObject. We need to avoid every extra byte.

**Comment by vexx32** [comment]
Don't _most_ `IEnumerable`s have a `Count` property anyway? Why is it lacking on here?

**Comment by mklement0** [comment]
@vexx32: [`IEnumerable`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerable) is about _lazy_ enumeration, so the count isn't known in advance - that's why the interface has no `.Count` property (applies to both the generic and the non-generic interface).

**Comment by SeeminglyScience** [comment]
Really you could just add

```csharp
public int Count => System.Linq.Enumerable.Count(this);
```

To `PSMemberInfoCollection<>` just to fix this scenario as it's pretty annoying to run into. Not many folks are likely to though.

**Comment by vexx32** [comment]
I'm down.

**Comment by iSazonov** [comment]
I believe the property store could be refactored as a whole (for performance).
/cc @powercode

**Comment by powercode** [comment]
@iSazonov Yes, I've two different branches on separate approaches for that. But it is non-trivial.

@vexx32 The operation of getting all properties for a type is expensive, since it has to look at many different sources - reflection, type adaptors, registered type info, instance members etc. So to expect a property Count is understandable but much harder to realize than it may seem.

Properties should by design guidelines be like fields, and not do excessive computations, so it may not be a good fit for this case.

**Comment by iSazonov** [comment]
@powercode You could share the code - somebody could help (Dongbo :-) ).

**Comment by powercode** [comment]
@iSazonov  Waste of his time so far.
